indexing-service/src/main/java/io/druid/indexing/common/task/CompactionTask.java

```diff
@@ -0,0 +1,266 @@
+      TaskToolbox toolbox,
+      String dataSource,
+      Interval interval
+  ) throws IOException, SegmentLoadingException
+  {
+    final List<DataSegment> usedSegments = toolbox
+        .getTaskActionClient()
+        .submit(new SegmentListUsedAction(dataSource, interval, null));
+    final Map<DataSegment, File> segmentFileMap = toolbox.fetchSegments(usedSegments);
+    final List<TimelineObjectHolder<String, DataSegment>> timelineSegments = CompactionTaskUtils.toTimelineSegments(
+        usedSegments,
+        interval
+    );
+    return Pair.of(segmentFileMap, timelineSegments);
+  }
+
+  private static DataSchema createDataSchema(
+      String dataSource,
+      Interval interval,
+      IndexIO indexIO,
+      ObjectMapper jsonMapper,
+      List<TimelineObjectHolder<String, DataSegment>> timelineSegments,
+      Map<DataSegment, File> segmentFileMap
+  )
+      throws IOException, SegmentLoadingException
+  {
+    // find metadata for interval
+    final List<QueryableIndex> segments = CompactionTaskUtils.loadSegments(timelineSegments, segmentFileMap, indexIO);
+    final Granularity expectedGranularity = segments.get(0).getMetadata().getQueryGranularity();
+    final boolean expectedRollup = segments.get(0).getMetadata().isRollup();
+
+    // check metadata
+    Preconditions.checkState(
+        segments.stream().allMatch(index -> index.getMetadata().getQueryGranularity().equals(expectedGranularity))
+    );
+    Preconditions.checkState(
+        segments.stream().allMatch(index -> index.getMetadata().isRollup() == expectedRollup)
+    );
+
+    // find merged aggregators
+    final List<AggregatorFactory[]> aggregatorFactories = segments.stream()
+                                                                  .map(index -> index.getMetadata().getAggregators())
+                                                                  .collect(Collectors.toList());
+    final AggregatorFactory[] mergedAggregators = AggregatorFactory.mergeAggregators(aggregatorFactories);
+    // conver to combining aggregators
+    final AggregatorFactory[] combiningAggregators;
+    if (mergedAggregators == null) {
+      combiningAggregators = null;
+    } else {
+      final List<AggregatorFactory> combiningAggregatorList = Arrays.stream(mergedAggregators)
+                                                                    .map(AggregatorFactory::getCombiningFactory)
+                                                                    .collect(Collectors.toList());
+      combiningAggregators = combiningAggregatorList.toArray(new AggregatorFactory[combiningAggregatorList.size()]);
+    }
+
+    // find granularity spec
+    final GranularitySpec granularitySpec = new UniformGranularitySpec(
+        new PeriodGranularity(interval.toPeriod(), null, null),
+        expectedGranularity,
+        expectedRollup,
+        ImmutableList.of(interval)
+    );
+
+    return new DataSchema(
+        dataSource,
+        ImmutableMap.of("type", "noop"),
```

Comment:
I think this will not work right with regard to numeric dimensions.

This data schema will essentially tell the index task to use metrics from combiningAggregators (which is good, assuming it's computed properly) and to auto-detect dimensions. But dimension auto-detection basically just treats everything that is not an input to an aggregator as a string. It won't retain the types they had in the original segment if it was a `long` or `float` dimension for example.

Reply:
Thank you for the good point. I thought about this and the only possible solution looks allowing user-defined dimensionsSpec in the compactionTask spec until we store data types of dimensions in somewhere. Does it make sense?

Reply:
Changed to accept an optional `dimensionsSpec`.

Reply:
We can do better than that, by examining what types the existing columns are. `storageAdapter.getColumnCapabilities(column).getType()` is the way to do that.